File: backend/src/main/java/com/hms/backend/service/MedicalRecordService.java
package com.hms.backend.service;

import com.hms.backend.entity.Appointment;
import com.hms.backend.entity.MedicalRecord;
import com.hms.backend.exception.ResourceNotFoundException;
import com.hms.backend.exception.DuplicateMedicalRecordException;
import com.hms.backend.repository.AppointmentRepository;
import com.hms.backend.repository.MedicalRecordRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class MedicalRecordService {

    @Autowired
    private MedicalRecordRepository medicalRecordRepository;

    @Autowired
    private AppointmentRepository appointmentRepository;

    public MedicalRecord saveMedicalRecord(MedicalRecord medicalRecord) {

        Long appointmentId = medicalRecord.getAppointment().getId();

        Appointment appointment = appointmentRepository.findById(appointmentId)
                .orElseThrow(() ->
                        new ResourceNotFoundException(
                                "Appointment not found with id: " + appointmentId));

        if (medicalRecordRepository.existsByAppointment(appointment)) {
            throw new DuplicateMedicalRecordException(
                    "Medical Record already exists for this appointment.");
        }

        medicalRecord.setAppointment(appointment);

        return medicalRecordRepository.save(medicalRecord);
    }

    public List<MedicalRecord> getAllMedicalRecords() {
        return medicalRecordRepository.findAll();
    }

    public MedicalRecord getMedicalRecordById(Long id) {
        return medicalRecordRepository.findById(id)
                .orElseThrow(() ->
                        new ResourceNotFoundException(
                                "Medical Record not found with id: " + id));
    }

    public MedicalRecord getMedicalRecordByAppointment(Long appointmentId) {

        Appointment appointment = appointmentRepository.findById(appointmentId)
                .orElseThrow(() ->
                        new ResourceNotFoundException(
                                "Appointment not found with id: " + appointmentId));

        return medicalRecordRepository.findByAppointment(appointment)
                .orElseThrow(() ->
                        new ResourceNotFoundException(
                                "Medical Record not found for Appointment ID: " + appointmentId));
    }

    public void deleteMedicalRecord(Long id) {

        MedicalRecord medicalRecord = medicalRecordRepository.findById(id)
                .orElseThrow(() ->
                        new ResourceNotFoundException(
                                "Medical Record not found with id: " + id));

        medicalRecordRepository.delete(medicalRecord);
    }
}

File: frontend/src/services/medicalRecordService.js
import api from "./api";

// =========================
// CREATE MEDICAL RECORD
// =========================
export const createMedicalRecord = async (medicalRecord) => {
    const response = await api.post("/medical-records", medicalRecord);

    return response.data;
};

// =========================
// GET ALL MEDICAL RECORDS
// =========================
export const getMedicalRecords = async () => {
    const response = await api.get("/medical-records");

    return response.data;
};

// =========================
// GET MEDICAL RECORD BY ID
// =========================
export const getMedicalRecordById = async (id) => {
    const response = await api.get(`/medical-records/${id}`);

    return response.data;
};

// =========================
// GET MEDICAL RECORD BY APPOINTMENT
// =========================
export const getMedicalRecordByAppointment = async (appointmentId) => {
    const response = await api.get(
        `/medical-records/appointment/${appointmentId}`
    );

    return response.data;
};

// =========================
// DELETE MEDICAL RECORD
// =========================
export const deleteMedicalRecord = async (id) => {
    const response = await api.delete(`/medical-records/${id}`);

    return response.data;
};